_posts/2019-08-18-Securing-Web-Applications-With-Keycloak.md

```diff
@@ -0,0 +1,608 @@
+For non-production purposes, it's easiest to just [download](https://www.keycloak.org/downloads.html) and run the standalone.
+At the time of writing, the latest release version was `6.0.1`.
+
+```bash
+$ curl https://downloads.jboss.org/keycloak/6.0.1/keycloak-6.0.1.zip --output keycloak-6.0.1.zip
+$ unzip keycloak-6.0.1.zip
+$ cd keycloak-6.0.1/bin/
+$ ./standalone
+```
+
+> For Windows users, there is also a `standalone.bat` in the same folder.
+
+The Keycloak server is now running on port 8080.
+
+> Use `-Djboss.socket.binding.port-offset` to change the port.  
+`-Djboss.socket.binding.port-offset=1000` will run the server on port 8080 + 1000 = 9080 
+
+Go to [http://localhost:8080](http://localhost:8080) and create an administrator account.
+You can now click on `Administration Console >` and log in using the account you've just created.  
+
+You are now on the pre-defined Master realm.
+
+> A realm manages a set of users, credentials, roles, and groups. 
+A user belongs to and logs into a realm and they are isolated from one another and can only manage and authenticate the users that they control.  
+
+The master realm is the highest level in the hierarchy. 
+Admin accounts in this realm have permissions to view and manage any other realm. 
+It's best to create a new realm to manage our application and users.
+
+## Creating a New Realm
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/create_realm.png" width="30%">
+</div>
+
+Create a new realm by clicking on the drop-down arrow next to the realm name in the upper left corner.
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/create_realm_2.png" width="60%">
+</div>
+
+In this example, heroes is chosen as the name of the realm.
+Feel free to change this to the name of your organisation if you have one.
+
+## Creating a Client
+
+Every application that interacts with Keycloak is considered to be a client.  
+Let's create one for the [Single Page App](https://en.wikipedia.org/wiki/Single-page_application) or SPA.  
+
+Look for the `Clients` tab in the menu and hit `Create`.  
+Pick a name you think is suitable and choose OpenID Connect (OIDC) as protocol.  
+The `Root URL` can remain blank.  
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/create_client.png" width="50%">
+</div>
+
+After saving, we can see all the configuration options of the client.  
+* `Valid Redirect URIs` should be set to `http://localhost:4200/*` as this is the address of our SPA  
+* Allow all origins for testing purposes
+* Single Page apps use a `public client` because they can not securely hide client credentials
+* Make sure the `Standard Flow` is enabled. All other flows can be disabled
+
+<div style="text-align: center;" >
+  <img src="/img/securing-web-applications-with-keycloak/configure_client.png" width="70%">
```

Comment:
Add an image with eg. `<img class="image fit" src="{{ '/img/devoxx-ma-2018/speakers-dinner.jpg' | prepend: site.baseurl }}" alt="Speakers Dinner" width="60%" />`, otherwise this won't scale good.